File: unittest/WholeBodyKinematics.cpp
#include <dwl/model/WholeBodyKinematics.h>
#define BOOST_TEST_MODULE DWL_TESTS
#include <boost/test/included/unit_test.hpp>



BOOST_AUTO_TEST_CASE(test_hyq) // specify a test case for hyq properties
{
    std::string urdf_file = DWL_SOURCE_DIR"/models/hyq.urdf";
    std::string yarf_file = DWL_SOURCE_DIR"/models/hyq.yarf";
    dwl::model::FloatingBaseSystem fbs;
    dwl::model::WholeBodyKinematics wkin;
    fbs.resetFromURDFFile(urdf_file, yarf_file);
    wkin.reset(fbs);


    // Creating an initial state for testing
    dwl::SE3 base_pos;
    dwl::Motion base_vel, base_acc;
    base_vel.setLinear(Eigen::Vector3d(1., 0., 0.));
    base_acc.setAngular(Eigen::Vector3d(0., 1., 0.));
    Eigen::VectorXd joint_pos = fbs.getDefaultPosture();
    Eigen::VectorXd joint_vel = Eigen::VectorXd::Zero(fbs.getJointDoF());
    Eigen::VectorXd joint_acc = Eigen::VectorXd::Zero(fbs.getJointDoF());


    // Checking the CoM position and velocity
    Eigen::Vector3d x, xd, xdd;
    wkin.computeCoMRate(x, xd, xdd,
						base_pos, joint_pos,
						base_vel, joint_vel,
						base_acc, joint_acc);
    BOOST_CHECK( x.isApprox(Eigen::Vector3d(-0.000278458, 0., -0.0460617), 1e-3) );
    BOOST_CHECK( xd.isApprox(Eigen::Vector3d(1., 0., 0.)) );


    // Checking the forward kinematics for the LF foot
    dwl::SE3Map pos_W =
        wkin.computePosition(base_pos, joint_pos,
                             fbs.getEndEffectorList());
    dwl::SE3 lf_pos(Eigen::Vector3d(0.370773, 0.324067, -0.57751),
    				Eigen::Vector4d(0.0928958, -0.364443, -0.0365662, 0.925859));
    BOOST_CHECK( pos_W.find("lf_foot")->second.getTranslation().isApprox(lf_pos.getTranslation(), 1e-3) );
    BOOST_CHECK( pos_W.find("lf_foot")->second.getQuaternion().isApprox(lf_pos.getQuaternion(), 1e-3) );


    // Checking the LF foot velocity
    dwl::MotionMap vel_W =
        wkin.computeVelocity(base_pos, joint_pos,
                             base_vel, joint_vel,
                             fbs.getEndEffectorList());
    dwl::Motion lf_vel(Eigen::Vector3d(1., 0., 0.), Eigen::Vector3d(0., 0., 0.));
    BOOST_CHECK( vel_W.find("lf_foot")->second.getLinear().isApprox(lf_vel.getLinear(), 1e-12) );
    BOOST_CHECK( vel_W.find("lf_foot")->second.getAngular().isApprox(lf_vel.getAngular(), 1e-12) );


    // Checking the LF foot acceleration
    dwl::MotionMap acc_W =
        wkin.computeAcceleration(base_pos, joint_pos,
                                 base_vel, joint_vel,
                                 base_acc, joint_acc,
                                 fbs.getEndEffectorList());
    dwl::Motion lf_acc(Eigen::Vector3d(-0.94828, 0., 0.206736), Eigen::Vector3d(0., 1., 0.));
    BOOST_CHECK( acc_W.find("lf_foot")->second.getLinear().isApprox(lf_acc.getLinear(), 1e-3) );
    BOOST_CHECK( acc_W.find("lf_foot")->second.getAngular().isApprox(lf_acc.getAngular(), 1e-3) );


    // Checking the fixed-based IK
    Eigen::VectorXd new_joint_pos = Eigen::VectorXd::Zero(fbs.getJointDoF());
    dwl::SE3Map pos_B = pos_W; // Note that we didn't move the base configuration
    wkin.computeJointPosition(new_joint_pos, pos_B);
    BOOST_CHECK( new_joint_pos.isApprox(joint_pos, 1e-12) );


    // Checking the joint velocity computation
    Eigen::VectorXd new_joint_vel = Eigen::VectorXd::Zero(fbs.getJointDoF());
    joint_vel.setRandom();
    dwl::MotionMap vel_B =
        wkin.computeVelocity(base_pos, joint_pos,
                             dwl::Motion(), joint_vel,
                             fbs.getEndEffectorList());
	wkin.computeJointVelocity(new_joint_vel, joint_pos, vel_B);
    BOOST_CHECK( new_joint_vel.isApprox(joint_vel, 1e-12) );


    // Checking the joint velocity computation
    Eigen::VectorXd new_joint_acc = Eigen::VectorXd::Zero(fbs.getJointDoF());
    joint_acc.setRandom();
    dwl::MotionMap acc_B =
        wkin.computeAcceleration(base_pos, joint_pos,
                                 dwl::Motion(), joint_vel,
                                 dwl::Motion(), joint_acc,
                                 fbs.getEndEffectorList());
	wkin.computeJointAcceleration(new_joint_acc, joint_pos, joint_vel, acc_B);
    BOOST_CHECK( new_joint_acc.isApprox(joint_acc, 1e-12) );
std::cout << joint_acc.transpose() << std::endl;
std::cout << new_joint_acc.transpose() << std::endl;
}

File: sample/python/WholeBodyKinematics.py
from __future__ import print_function # python3-style print()
import os
import dwl
import numpy as np


# Configure the printing
np.set_printoptions(suppress=True, linewidth=1000)

# Construct an instance of the WholeBodyKinematics class, which wraps the C++ class.
ws = dwl.WholeBodyState()
fbs = dwl.FloatingBaseSystem()
wkin = dwl.WholeBodyKinematics()

# Resetting the system from the hyq urdf file
fpath = str(os.path.dirname(os.path.abspath(__file__)))
fbs.resetFromURDFFile(fpath + "/../../models/hyq.urdf",
                      fpath + "/../../models/hyq.yarf")
wkin.reset(fbs)

# Define the DoF after initializing the robot model
ws.setJointDoF(fbs.getJointDoF())


# The robot state
ws.setBaseSE3(dwl.SE3_RPY(np.zeros(3),
                          np.array([0., 0., 0.])))
ws.setBaseVelocity_W(dwl.Motion(np.array([0., 0., 0.]),
                                np.array([0., 0., 0.])))
ws.setBaseAcceleration_W(dwl.Motion(np.array([0., 0., 0.]),
                                    np.array([0., 0., 0.])))
ws.setJointPosition(0.75, fbs.getJointId("lf_hfe_joint"))
ws.setJointPosition(-1.5, fbs.getJointId("lf_kfe_joint"))
ws.setJointPosition(-0.75, fbs.getJointId("lh_hfe_joint"))
ws.setJointPosition(1.5, fbs.getJointId("lh_kfe_joint"))
ws.setJointPosition(0.75, fbs.getJointId("rf_hfe_joint"))
ws.setJointPosition(-1.5, fbs.getJointId("rf_kfe_joint"))
ws.setJointPosition(-0.75, fbs.getJointId("rh_hfe_joint"))
ws.setJointPosition(1.5, fbs.getJointId("rh_kfe_joint"))

ws.setJointVelocity(0.2, fbs.getJointId("lf_haa_joint"))
ws.setJointVelocity(0.75, fbs.getJointId("lf_hfe_joint"))
ws.setJointVelocity(1., fbs.getJointId("lf_kfe_joint"))

ws.setJointAcceleration(0.2, fbs.getJointId("lf_haa_joint"))
ws.setJointAcceleration(0.75, fbs.getJointId("lf_hfe_joint"))
ws.setJointAcceleration(1., fbs.getJointId("lf_kfe_joint"))




# Computing the jacobians
print()
print('Stack of Jacobians:')
jacobian = wkin.computeJacobian(ws.getBaseSE3(), ws.getJointPosition(),
                                fbs.getEndEffectorList(dwl.FOOT))
print(' ', jacobian)



 
print()
print('Floating-base Jacobian (LF foot):')
floating_jac = np.zeros([6, 6])
wkin.getFloatingBaseJacobian(floating_jac, jacobian['lf_foot']);
print(' ', floating_jac)
  
print()
print('Fixed Jacobian (LF foot):')
fixed_jac = np.zeros([6, fbs.getJointDoF()]) 
wkin.getFixedBaseJacobian(fixed_jac, jacobian['lf_foot']);
print(' ', fixed_jac)
 
 
 
 
 
# Computing the contact positions
print()
print('Contact SE3:')
contact_pos_W = wkin.computePosition(ws.getBaseSE3(), ws.getJointPosition(),
                                     fbs.getEndEffectorList(dwl.FOOT))
print(' ',contact_pos_W.asdict()['lf_foot'])
ws.setContactSE3_W(contact_pos_W)
 
 
 
 
# Computing the contact velocities
print()
print('Contact velocity:')
contact_vel_W = wkin.computeVelocity(ws.getBaseSE3(), ws.getJointPosition(),
                                     ws.getBaseVelocity_W(), ws.getJointVelocity(),
                                     fbs.getEndEffectorList(dwl.FOOT))
print(' ',contact_vel_W.asdict()['lf_foot'])
ws.setContactVelocity_W(contact_vel_W);
 
 
 
 
# Computing the contact accelerations
print()
print('Contact acceleration:')
contact_acc_W = wkin.computeAcceleration(ws.getBaseSE3(), ws.getJointPosition(),
                                         ws.getBaseVelocity_W(), ws.getJointVelocity(),
                                         ws.getBaseAcceleration_W(), ws.getJointAcceleration(),
                                         fbs.getEndEffectorList(dwl.FOOT))
print(' ', contact_acc_W.asdict()['lf_foot'])
 
 
 
 
# Computing the Jdot *qdot of the contacts
print()
print('Contact Jd*qd:')
contact_jdqd_W = wkin.computeJdQd(ws.getBaseSE3(), ws.getJointPosition(),
                                  ws.getBaseVelocity_W(), ws.getJointVelocity(),
                                  fbs.getEndEffectorList(dwl.FOOT))
print(' ', contact_jdqd_W.asdict())
 
 
 
 
# Computing the joint positions
wkin.setIKSolver(1.0e-12, 50)
joint_pos = np.zeros(fbs.getJointDoF())
contact_pos_B = dwl.SE3Map()
contact_pos_B['lf_foot'] = dwl.SE3(np.array([0.371,0.207,-0.589]), np.eye(3))
contact_pos_B['lh_foot'] = dwl.SE3(np.array([-0.371,0.207,-0.589]), np.eye(3))
joint_pos_init = fbs.getDefaultPosture()
print()
print('Joint position:')
if wkin.computeJointPosition(joint_pos,
                             contact_pos_B,
                             joint_pos_init):
    print(' ', joint_pos.transpose())
else:
    print(' The IK problem could not be solved')
 
 
 
 
# Computing the joint velocities. Note that you can create and dwl.MotioMap()
# object, similar as before
joint_vel = np.zeros(fbs.getJointDoF())
print()
print('Joint velocity:')
contact_vel_B = dwl.MotionMap()
contact_vel_B['lf_foot'] = dwl.Motion(np.zeros((3,1)), np.matrix([ [1.],[0.],[0.] ]))
contact_vel_B['lh_foot'] = dwl.Motion(np.zeros((3,1)), np.matrix([ [0.],[1.],[0.] ]))
contact_vel_B['rf_foot'] = dwl.Motion(np.zeros((3,1)), np.matrix([ [0.],[0.],[1.] ]))
wkin.computeJointVelocity(joint_vel,
                          ws.getJointPosition(),
                          contact_vel_B)
print(' ', joint_vel.transpose())
 
 
 
 
# Computing the joint accelerations
joint_acc = np.zeros(fbs.getJointDoF())
print()
print('Joint acceleration:')
contact_acc_B = contact_acc_W
wkin.computeJointAcceleration(joint_acc,
                              ws.getJointPosition(),
                              ws.getJointVelocity(),
                              contact_acc_B)
print(' ', joint_acc.transpose())
 
 
 
 
# Computing the CoM position, velocity and acceleration
print()
print('CoM position:')
print(' ', wkin.computeCoM(ws.getBaseSE3(), ws.getJointPosition()).transpose())
c_pos = np.zeros(3)
c_vel = np.zeros(3)
c_acc = np.zeros(3)
wkin.computeCoMRate(c_pos, c_vel, c_acc,
                    ws.getBaseSE3(), ws.getJointPosition(),
                    ws.getBaseVelocity_W(), ws.getJointVelocity(),
                    ws.getBaseAcceleration_W(), ws.getJointAcceleration())
print('CoM velocity:')
print(' ', c_vel)
print('CoM acceleration:')
print(' ', c_acc)
 
 
 
 
print()
print('Constrained acceleration')
ws.setBaseAcceleration_W(dwl.Motion(np.matrix([ [1.],[0.],[0.] ]),
                                    np.matrix([ [0.],[0.],[0.] ])))
joint_facc = np.zeros((fbs.getJointDoF(),1))
wkin.computeConstrainedJointAcceleration(joint_facc,
                                         ws.getBaseSE3(), ws.getJointPosition(),
                                         ws.getBaseVelocity_W(), ws.getJointVelocity(),
                                         ws.getBaseAcceleration_W(),
                                         fbs.getEndEffectorList(dwl.FOOT))
print(' joint acceleration:', joint_facc.transpose())